pkg/tfcli/options.go

```diff
@@ -220,16 +202,24 @@ func (r Resource) GetAddress() string {
 	return fmt.Sprintf(fmtResourceAddress, r.LabelType, r.LabelName)
 }
 
+// ProviderRequirement holds values for the Terraform HCL provider requirements
+type ProviderRequirement struct {
+	Source  string
+	Version string
+}
+
+// ProviderConfiguration holds the provider configuration body
+type ProviderConfiguration []byte
```

Comment:
```suggestion
type ProviderConfiguration struct {
	Source  string
	Version string
	Credentials map[string]interface{}
}
```

If we pass around this single object, we'd be leaking less information about where it ends up. From Crossplane provider developer perspective, they don't need to know whether it's called requirement or configuration. We can hide this until go template section.

While I was working with terraform state stuff, I find it more handy to pass around `map[string]interface{}` instead of `[]data` so that we can make changes more easily down the line.

Reply:
I am not sure. I understand the intent of hiding our implementation details but I believe by following similar naming it would be easier for developers to clearly see which struct expects what:

https://www.terraform.io/docs/language/providers/requirements.html
https://www.terraform.io/docs/language/providers/configuration.html

Also `Credentials` is not correct here, because it could contains other information in [provider configurations](https://www.terraform.io/docs/language/providers/configuration.html) which are usually not credentials (e.g. region)

Reply:
But as a simplification (for passing around a single object) I would be ok with the following:

```
type ProviderConfiguration map[string]interface{}

type ProviderSetup struct {
	ProviderRequirement
        ProviderConfiguration
}
```

Reply:
@muvaf wdyt?

Reply:
Looks good to me.

Reply:
@turkenh we need[ `TERRAFORM_VERSION`](https://github.com/crossplane-contrib/terrajet/pull/72/files#diff-68d8aa09b23aabebdc356ea5a16ebfa926b18b454366ac4c50f43ab8e8429178R82) as well during the reconstruction of the tfstate. `ProviderSetup` you suggested could be a good candidate to have this as field.

Reply:
Ok, I'll add it.

Reply:
Hmm, TerraformVersion didn't look great under `ProviderSetup` 🤔 

```
type ProviderSetup struct {
	TerraformVersion string
	Requirement      ProviderRequirement
	Configuration    ProviderConfiguration
}
``` 

Wondering if the following would be better: 

```
type TerraformSetup struct {
	Version          string
	Requirement      ProviderRequirement
	Configuration    ProviderConfiguration
}
```

Reply:
Going with the latter.